### ISSUE TITLE
EasyRSA on Windows 11 with `mksh` hangs

### ISSUE THREAD

**Issue by TinCanTech** [OpenVPN/easy-rsa#1075]
#### Follow-up set: #1078

-----

This is specifically:
- **Windows 11**
  Other versions of Windows may also be affected. If you are able to test another version of Windows which experiences similar _hangs_ then please leave your feedback here.
- `mksh` built-in command `read`
- The hang is caused by `easyrsa` use of `read` to ask for user input.
- Easy-RSA tool chain is ancient and this could be _the last straw_. 

Required testing:
- **First test**: On Windows 11, open `sh.exe` and test the behavior of `read`.

Volunteers welcome.

If you have access to any version of Windows then you can help by testing.

**Comment by TinCanTech** [comment]
First test:

On Windows 11, open `sh.exe`.

To test command `read`, simply enter the word `read` into the terminal and press `[enter]`. Then press `[enter]` again, and you should return to the prompt.

This MUST be fixed -- There is no work-around.

**Comment by TinCanTech** [comment]
**Requirements**: Windows 11 with `git-for-windows`, installed. Application for `7zip` file format decompression.

Instructions for testing `busybox.exe`:

`git` setup and download `TinCanTech-upload-busybox`:
- `git clone https://github.com/Openvpn/easy-rsa.git easyrsa-busybox`
- `cd easyrsa-busybox`
- `git checkout -b TinCanTech-upload-busybox`
- `git pull --no-ff --no-commit https://github.com/TinCanTech/easy-rsa.git upload-busybox`

Install `busybox.exe` for use:
- Extract all files from `distro/dev/busybox-v4.7z into `distro/dev` folder.
- Copy `busybox.exe` to `C:\Program Files\Openvpn\easy-rsa\bin`

Change your current Windows Easy-RSA installation to use `busybox.exe`:
- Edit: `C:\Program Files\Openvpn\easy-rsa\bin\easyrsa-shell-init.sh`
- Change the last line `#159` and over-write on save:
  From: `sh.exe`
  To: `busybox.exe`

Start your current Windows Easy-RSA installation from `EasyRSA-Start.bat`.

**Comment by TinCanTech** [comment]
Follow-up set-to: #1076

**Comment by TinCanTech** [comment]
@OutOfEspresso Hi,

Busy-box for Windows would be my preferred solution because we can build it ourselves. However, OpenVPN developers feel that this is another unnecessary dependency.

If we have a new version of MKSH then I can provide a .zip file for Windows testing. All I need is a link to the MKSH project.

**Comment by TinCanTech** [comment]
@OutOfEspresso Sorry but your random binary is **not** suitable for EasyRSA.

**Comment by TinCanTech** [comment]
I am not in the habit of running random binaries from strangers on the internet and this is no exception.

If you are serious then do what you need to do.

**Comment by avih** [comment]
> The port was done since we had to replace the commercial "MKS Toolkit" with a free solution. We did not find anything suitable and thus, decided to do a port of MKSH our own and give the result back to the community as a one shot. This shot was beta14, which had to be released before 2013-06-30 due to legal reasons.

Who is "we", and what's the context?

Is the current `sh.exe` which is `R39-w32-beta14` in the current release the same binary which you or your colleagues compiled back then?

**Comment by avih** [comment]
Thanks for the info.

So basically this port was developed internally in Salaris AG for internal reasons (replace MSK toolkit), and together you decided to contribute the source patches and binaries back to the mksh project, which was then published at http://www.mirbsd.org/permalinks/wlog-10_e20130718-tg.htm#e20130718-tg_wlog-10  and later the EasyRSA people probably found that binary on that page and decided to use it?

> And yes, the binary EasyRSA provides is identical to the one contained in this Zip

Yes, I can confirm that. Both have CRC32 of `e09ed595` and md5sum of `66ff1278e852dedd3b005a54483cab42` .

So that's a pretty good start. At least now we have the sources (which need VS2008, and don't include `nedmalloc_v105_svn1078.zip`, but that file is still available on soureceforge), and presumably it would be possible to produce a binary.

> After the product requiring it was depreciated, the only remaining usecase was my personal use. As mentioned, i made some fixes until 2017 and thats it for now.

This sounds to me like you worked on the public mksh-w32 sources for your personal needs, so I don't see how anyone could have an ownership  claim on that development, other than yourself, right?

In that case, could you please publish the sources you have which were used to compile that last version from 2017? For instance at a github repo you'll create for it? (you could tag the 2013 sources as beta14, then overwrite the files with the beta 28 sources, and tag it as such).

I don't see a reason why `nedmalloc_v105_svn1078.zip` can't be included too. As far as I can tell it has the Boost license (so that license would have to be included too, which it is in that zip ffile).

> mksh-w32(at)gmx.net address, mentioned in the ReadMe was removed too. Since there was no incoming mail for lots of years

I guess it just worked till this hang issue with windows 11...

**Comment by avih** [comment]
> @OutOfEspresso Sorry but your random binary is **not** suitable for EasyRSA.

You say that, but that's exactly what `distro/windows/bin/sh.exe` is, which apparently suites EasyRSA quite well for the past 11 years.

It's the same binary from http://www.mirbsd.org/permalinks/wlog-10_e20130718-tg.htm#e20130718-tg_wlog-10 , and which the mksh project published as is without building it from the source themselves, and the EasyRSA project did exactly the same and used this binary without building it themselves, or else I'm quite sure the md5sum would have changed.

So basically the EasyRSA project trusted the mksh people, who apparently trusted "Michael Langguth and Scalaris AG" and publish their binary unmodified.

That's not at all to say that this beta28 binary should be taken without sources - it shouldn't IMO. Even if the sources do exist, then it would still be unwise to take an unknown binary.

But at least now we presumably have the sources to `distro/windows/bin/sh.exe`, and with some effort someone might be able to build them on a modern system, and maybe debug this win11 hang.

And if the beta28 files would become available as well, then those might help too.

**Comment by TinCanTech** [comment]
> > @OutOfEspresso Sorry but your random binary is **not** suitable for EasyRSA.
> 
> You say that, but that's exactly what `distro/windows/bin/sh.exe` is

> I attached the latest version just here mksh-w32-beta28.zip[URL Redacted]. Good luck :-)

Which is **not** the same version distributed by Easy-RSA.

@OutOfEspresso Thanks for the details you have provided, perhaps you could have done that initially.

@avih Thanks for pushing this.

Once again, I do not think the new binary is suitable under the current circumstances.

**Comment by avih** [comment]
> > I attached the latest version just here mksh-w32-beta28.zip[URL Redacted]. Good luck :-)
> 
> Which is **not** the same version distributed by Easy-RSA.

I did not say that. I said the current `sh.exe` in the distro is exactly such "random binary", presumably by @OutOfEspresso, which EastRSA didn't compile themselves, and the mksh project also didn't compile themselves and instead published it based on trust that it's OK.

I would have definitely expected EasyRSA to compile it themselves and not trust a random binary which even the mksh people say they didn't compile themselves.

But that's water under the bridge now.

> I do not think the new binary is suitable under the current circumstances.

Agreed.

Which is why I think that as a starting point, EasyRSA should make an effort to build `sh.exe` from the available sources of beta-14 at the github CI. It might not be easy, but it should be possible, either to compile it using new VS, or maybe adapt it to mingw.

And if the beta-28 sources become available, check whether those can help too.

**Comment by TinCanTech** [comment]
> I said the current `sh.exe` in the distro is exactly such "random binary"

No, it is not.

**Comment by avih** [comment]
> No, it is not.

Well, it has the s

> > I said the current `sh.exe` in the distro is exactly such "random binary"
> 
> No, it is not.

It has the same md5sum as the binary available at http://www.mirbsd.org/permalinks/wlog-10_e20130718-tg.htm#e20130718-tg_wlog-10 and which the mksh people claim they didn't compile themselves and instead publish it because someone sent it to them.

How would you call that anything other than "random binary which we didn't compile ourselves and we don't know who compiled it or from which sources" ?

**Comment by TinCanTech** [comment]
The binary offered by @OutOfEspresso came with **zero** verifiable details.

The binaries shipped by Easy-RSA have all been verified.

Your line of argument is not productive.

**Comment by avih** [comment]
> Meanwhile I agree'd with the MirOS Project, that we most likely will release it at the same location as beta 14.

Sounds good.

Meanwhile, I've compiled beta 14, and I suspect that the sources are not of the binary, because as far as I can tell there's a leftover bug at the code which strictly prevent compilation.

Here's my patch to make it compile. Care to comment if I'm missing something?

<details>
<summary>patch to comment-out mtime() in liblan/systools.c</summary>

```patch
commit baabf4e2aee04ce39c3ec23d2189d487bd557731
Author: Avi Halachmi (:avih) <avihpit@yahoo.com>
Date:   Tue Aug 6 20:02:07 2024 +0300

    liblan: comment broken and unused mtime()
    
    The file liblan/systools.c is required for the build (for "sleep",
    and more), however, it doesn't compile with WIN32 defined, because
    "timezone" in mtime() is unresolved (it looks like a WIP bug), and
    it doesn't compile without WIN32 defined because neither msvc nor
    mingw have <sys/statvfs.h>.
    
    However, this function is unused by the code, so just comment it out.
    
    However also, this might be an indication that the binary mksh.exe in
    mksh-w32-beta14.zip might be built not from this exact source.
    
    The project can now be built using VS 2015 cl.exe (32 bit):
    - delete mksh/setmode.c (it's unused).
    - cl /Femksh.exe /D WIN32 /D LIBLAN /I liblan /I nedmalloc
         liblan/*.c mksh/*.c user32.lib winmm.lib advapi32.lib

diff --git a/liblan/systools.c b/liblan/systools.c
index a3b09bb..ca2d610 100644
--- a/liblan/systools.c
+++ b/liblan/systools.c
@@ -104,6 +104,11 @@ void msleep(long ms)
 /*
  *	get the system time in ms since utc 1970
  */
+/*
+ * mtime() doesn't compile with WIN32 because "timezone" is unresolved.
+ * however, it's unused by mksh, so disable it for now
+ */
+/*
 uint64 mtime(void)
 {
 	uint64	ret;
@@ -134,6 +139,7 @@ uint64 mtime(void)
 
 	return(ret);
 }
+*/
 
 /*
  *	unique interface to statfs like information
```
</details>

And here's the `mtime` function in liblan/systools.c. Note how `timezone` is not defined anywhere (also not outside that code):
```c
uint64 mtime(void)
{
	uint64	ret;

#ifdef WIN32
	SYSTEMTIME	st;
	struct tm	tm;

	GetSystemTime(&st);

	tm.tm_year  = st.wYear - 1900;
	tm.tm_mon   = st.wMonth - 1;
	tm.tm_mday  = st.wDay;
	tm.tm_hour  = st.wHour;
	tm.tm_min   = st.wMinute;
	tm.tm_sec   = st.wSecond;
	tm.tm_isdst = 0;

	ret = (uint64)mktime(&tm)*1000;
	ret += (timezone * -1000) + st.wMilliseconds;
#else
	struct timeb	tb;

	ftime(&tb);

	ret= (uint64)tb.time*1000 + tb.millitm;
#endif

	return(ret);
}
```

**Comment by avih** [comment]
> From my "C:\Program Files (x86)\Visual Studio 9.0\VC\include\time.h":
> `_CRT_INSECURE_DEPRECATE_GLOBALS(_get_timezone) _CRTIMP extern long timezone;`

Interesting.

I used msvc 2015 (but sort of portable thing, extracted from a full install, so I don't have the IDE etc, just the build tools and the SDK).

In `time.h` there's `_timezone` (and `struct timezone`) but not `long timezone`.
FWIW, in mingw time.h it's the same (no `timezone`).

However, in MSVC 6 there is indeed `timezone` in `time.h`, but there were too many errors when trying to compile ksh.

So either 2013 is the last which has `timezone`, or my msvc 2015 SDK is lacking stuff.
But anyway, that was the only blocker, and `mtime` is unused, so I just commented it out. Other than this, by default there were only 3 printf format warnings at `mksh/win32.c`.

> Forgot to mention: Cool! As i see it, you are the second one, succsessfully compiling mksh.exe on the planet. :-)

I guess. It doesn't look like the EasyRSA people compiled it..
